Loadbalance

shellshocked2003 opened the pull request:
Addresses issues with load balance among different processes. This branch implements:
* Load balancer that automatically divides input rock mass into approximately equal volumes to maintain similar loads across nodes
* Removal of "processor" joints to restore actual rock mass once all joints have been processed

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/Block.scala:
```diff
@@ -28,12 +47,50 @@ case class Face(normalVec: (Double,Double,Double), distance: Double, phi: Double
+    (math.abs(block1.centerX - updatedBlock2.centerX) < tolerance) &&
+    (math.abs(block1.centerY - updatedBlock2.centerY) < tolerance) &&
+    (math.abs(block1.centerZ - updatedBlock2.centerZ) < tolerance) &&
+    (!faceMatches.contains(false))
```
How about getting rid of the `map` on line 84 and instead just doing something like `faces forall { case (f1, f2) => Face.compareFaces(f1, f2)` ?

Seems more Scala-esque to me, and we have similar code in other places I think.

shellshocked2003 replied:
Makes sense, I changed this.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/Block.scala:
```diff
@@ -28,12 +47,50 @@ case class Face(normalVec: (Double,Double,Double), distance: Double, phi: Double
+    * @param block2 Second input block
+    * @return True if blocks are equal, otherwise false
+    */
+  def compareBlocks(block1: Block, block2: Block, tolerance: Double=NumericUtils.EPSILON):
```
I think we should implement this functionality as part of the `equals` method for `Block` rather than as part of the companion object. The same goes for the `Face` comparison code above.

In Java and Scala, the `equals` method is treated specially, and implementing it well can help our code play nicely with tools like the built-in collections.

So what I'm saying is that our code will look more like `face1 == face2` rather than `Face.compareFaces(face1, face2)`.

shellshocked2003 replied:
Yeah, I agree. Thought about doing this, but was in a rush to get something working. Haven't done this before, but from what I can find online it looks like you also need to define a `hashcode` method. Not familiar with that, but I can look into it.

jhkolb replied:
Yeah, and actually I'm having second thoughts because we're not checking for _exact_ equality, which is what the `equals` method is normally for. What about defining an `approximateEquals` method for `Block` objects? What do you think about that idea?

shellshocked2003 replied:
That seems fine to me. So, you're just requesting a name change to the method then?

jhkolb replied:
Actually, what I was suggesting was that you move this from the companion object into the class itself. So instead of `Block.compareBlocks(block1, block2)` the code would look more like `block1.compareTo(block2)`.

Up to you if we go ahead and do that, but yeah, at least a name change to something like `approximateEquals` is a good idea.

shellshocked2003 replied:
Got it. I'll change it; what you're suggesting makes more sense.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/Block.scala:
```diff
@@ -335,8 +392,9 @@ case class Block(center: (Double,Double,Double), faces: Seq[Face]) {
-        if (face.c < -NumericUtils.EPSILON) {
+        // If normal is -z-axis, order needs to be reversed to maintain clockwise orientation
+        // rotation matrix is identity matrix in this case
+        if (math.abs(face.c + 1.0) < NumericUtils.EPSILON) {
```
What's up with this change? Don't we still need to do something if `-1.0 < face.c < -NumericUtils.EPSILON` ?

shellshocked2003 replied:
The rotation matrix actually takes care of all cases except when the normal vectors are parallel, so it's not necessary to do anything unless the normal vectors are parallel but point in opposite directions.

jhkolb replied:
Great! Do you think you could add a short comment about that, just for those of us who aren't as geometrically inclined? :)

shellshocked2003 replied:
Added some comments here and in `Block.rotationMatrix` as well.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/Block.scala:
```diff
@@ -396,12 +454,38 @@ case class Block(center: (Double,Double,Double), faces: Seq[Face]) {
   }
 
   /**
+    * Calculates the volume of the block
```
Could you provide a reference to the method you're using to compute volume here in the comments? Something as simple as a link would suffice.

shellshocked2003 replied:
Added a link to a pdf that gives a pretty good description on how to do this.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/Block.scala:
```diff
@@ -396,12 +454,38 @@ case class Block(center: (Double,Double,Double), faces: Seq[Face]) {
+      }
+    }
+
+    volIncrements.fold(0.0) (_ + _)/6.0
```
It's awesome that you're being functional and using `fold`! But... can't we just use the more clear `sum` method here?

shellshocked2003 replied:
Haha! Sure, that would seem to make more sense.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/CommandReader.scala:
```diff
@@ -23,7 +23,7 @@ object CommandReader {
       c.copy(numSeedJoints = x)
     } validate { x =>
-      if (x > 0) success else failure("Number of seed joints must be positive")
+      if (x >= 0) success else failure("Number of seed joints must be greater than or equal to 0")
```
Aha, good catch. I guess someone could want to avoid using any seed joints.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/Delaunay.scala:
```diff
@@ -62,7 +62,7 @@ object Delaunay {
     */
     
-    val EPSILON = 0.0000001
+    val EPSILON = NumericUtils.EPSILON
```
Instead of just changing the definition of `EPSILON` here, why don't we just change all references to `EPSILON` to references to `NumericUtils.EPSILON` in the code below?

shellshocked2003 replied:
Good point.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/Joint.scala:
```diff
@@ -204,12 +204,15 @@ object Joint {
 case class Joint(normalVec: (Double, Double, Double), localOrigin: (Double, Double, Double),
                  center: (Double, Double, Double), phi: Double, cohesion: Double,
-                 shape: Seq[((Double, Double, Double),Double)], dipAngleParam: Option[Double]=None,
+                 shape: List[((Double, Double, Double),Double)], dipAngleParam: Option[Double]=None,
```
Why is this changed to `List`? That was what we had originally, but then I went back and changed things to `Seq` a while ago because it appears to be the recommended practice that we use the `Seq` interface and let Scala pick a concrete implementation.

Is there something we need to do that requires a `List`? I wonder if we couldn't a way to achieve the same thing with `Seq` after doing some research.

shellshocked2003 replied:
Yeah, I remember that. The reason for the change is related to Spark. If left as `Seq`, the compiler for some reason picks `Stream` which is non-serializable. Spark only works with serializable objects, so that's why I changed it. Even explicitly declaring `Joint` as `extends Serializable` doesn't fix the issue.

jhkolb replied:
Ahh, right I remember some issues with this before too. Okay, we can use `List`, and perhaps someday when we have more time we can revisit this.

shellshocked2003 replied:
Sounds good to me.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+    *                      Should be one less than number of cores used.
+    * @return Set of processor joints that will be used to seed the initial RDD
+    */
+  def generateSeedJoints(rockVolume: Block, numSeedJoints: Integer):
```
Hmm, I think we should try to be consistent on our terminology of seed joints vs processor joints.

In fact, do we even need our old notion of seed joints now? We could just use processor joints to divide up the work automatically, right? That would definitely simplify things from the user's end.

And perhaps now instead of having the number of seed joints as the second parameter, we want the number of equal-volume partitions to generate for the rock volume.

shellshocked2003 replied:
Got rid of the whole seed joint thing across all files. The user now specifies the number of processors that are being used in the analysis and the load balance is done based on that.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+                         Seq[Joint] = {
+    // Calculate bounding box
+    val vertices = rockVolume.findVertices.values.flatten
+    val x_max = (vertices.map{ case vertex => vertex._1}).max
```
How about `vertices.map(_._1)` and so on for the other extrema instead of the more complex `case` stuff here?

shellshocked2003 replied:
Sorry, got caught up in the syntax instead of thinking what the hell I was actually doing.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+    val y_min = (vertices.map{ case vertex => vertex._2}).min
+    val z_max = (vertices.map{ case vertex => vertex._3}).max
+    val z_min = (vertices.map{ case vertex => vertex._3}).min
+    val boundingBox = (x_min, y_min, z_min, x_max, y_max, z_max)
```
This is getting picky, but I wonder if a bit more structure is in order here, i.e. defining `boundingBox` as a pair of 3-tuples, one for the minima and another for the maxima, or as a triple of pairs for the x, y, and z minimum and maximum.

shellshocked2003 replied:
I agree, but for the sake of time maybe we can leave it as is for now?

jhkolb replied:
Sure, that sounds reasonable.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+    val centroid = rockVolume.centroid
+    val centroidVolume = Block(centroid, rockVolume.updateFaces(centroid))
+    val seedJoints =
+      findProcessorJoints(Seq[Joint](), (centroidVolume.centerX, centroidVolume.centerY, centroidVolume.centerZ),
```
For the first argument, how about `Seq.empty[Joint]`? It's just a bit more explicit, but not really a big deal.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+                                  Seq[Joint] = {
+    // Calculate bounding box
+    val vertices = initialVolume.findVertices.values.flatten
+    val x_max = (vertices.map{ case vertex => vertex._1}).max
```
Like before, I think we can do `map(_._1)` and similar for the other extrema instead of using `case`.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+    }
+    val sortedBlocks = nonRedundantBlocks.sortWith{ (left, right) =>
+      centroidCompare(left.centroid, right.centroid) }
+    val tolerance = 0.05
```
Do we want to make `tolerance` a parameter with a sensible default value?

shellshocked2003 replied:
I was debating about that, but I don't think we should. If the value is too small, it can cause some issues with convergence since the way in which volume varies can be quite erratic depending on the shape of the input volume. Once we have a better handle on how our floating point errors are getting generated we could make it a parameter?

jhkolb replied:
Sure, I think you have a good point. Would you mind making it a constant defined at the beginning of the class, i.e. write something like `val VOLUME_TOLERANCE = 0.5` ? That way we can easily keep track of it and make changes in the future.

shellshocked2003 replied:
Yeah, that makes sense.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+        joint +: joints
+      } else {
+        // If last block isn't small enough
+        val remainingBlock = sortedBlocks.tail.head
```
What about just doing `sortedBlocks(1)` here?

shellshocked2003 replied:
Figured it would fail more gracefully this way. It there anything bad about doing it this way?

jhkolb replied:
Oh, wouldn't `val remainingBlock = sortedBlocks(1)` have the same failure behavior as your current code? I just think it's more readable and makes it clear that you're grabbing the second element.

shellshocked2003 replied:
The result from `sortedBlocks(1)` is that it tells you index out of bounds while `sortedBlocks.tail.head` tells you it's an empty list. I guess it doesn't really matter, I can change it if you like?

jhkolb replied:
Okay, I'll leave it up to you :)

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+      val centerVec0 = DenseVector[Double](center._1, center._2, center._3)
+      val centerVec1 = DenseVector[Double](boundingBox._4, boundingBox._5, boundingBox._6)
+      // First distance guess
+      val x_diff0 =centerVec0(0) - boundingBox._1
```
Missing a space after `=` on these lines.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+    * @param desiredVolume Desired volume for each piece cut by processor joints
+    * @return List of processor joints that will be used to seed initial RDD
+    */
+  @tailrec
```
This is nice.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+      centroidCompare(left.centroid, right.centroid) }
+
+    // Secant step
+    val a = Seq(initialDist0, initialDist1).min
```
It seems a little weird to make these two-element `Seq`s. What about something like

``` scala
val (a,b) = if (initialDist0 < initialDist1) {
    (initialDist0, initialDist1)
} else {
    (initialDist1, initialDist0)
}
```

Not sure if the syntax is perfect, but you get the idea.

shellshocked2003 replied:
Yeah, this is much clearer.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+    // Secant step
+    val a = Seq(initialDist0, initialDist1).min
+    val b = Seq(initialDist0, initialDist1).max
+    val f_a = if (initialDist1 > initialDist0) {
```
You could do something similar to my comment above to set `f_a` and `f_b` at the same time.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+    // Bisection step
+    val bisectionDist = (initialDist0 + initialDist1)/2.0
+    // Select new step
+    val newDist = if ((secantDist < a) || (secantDist > b) || (secantDist.isNaN)) bisectionDist
```
Add some braces around the branches here to conform to our coding style.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+          if ((x_incr < 0.0) &&
+              (y_incr < 0.0) &&
+              (z_incr < 0.0))
+            -math.sqrt(x_incr*x_incr + y_incr*y_incr + z_incr*z_incr)
```
Braces around the branches here.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+          if ((x_incr < 0.0) &&
+              (y_incr < 0.0) &&
+              (z_incr < 0.0))
+            -math.sqrt(x_incr*x_incr + y_incr*y_incr + z_incr*z_incr)
```
Braces.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+          if ((x_incr < 0.0) &&
+              (y_incr < 0.0) &&
+              (z_incr < 0.0))
+            -math.sqrt(x_incr*x_incr + y_incr*y_incr + z_incr*z_incr)
```
Braces

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+          if ((x_incr < 0.0) &&
+              (y_incr < 0.0) &&
+              (z_incr < 0.0))
+            -math.sqrt(x_incr*x_incr + y_incr*y_incr + z_incr*z_incr)
```
Braces

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+    *                   maximum number of iterations have not been exceeded
+    */
+  @tailrec
+  private def secantBisectionSolver(initialDist0: Double, initialDist1: Double, block: Block, 
```
I'm nothing close to a numerical methods expert, so is the term "secant bisection solver" meaningful in this context? Otherwise we may to include some kind of reference in the comments.

shellshocked2003 replied:
Yeah, I think it's meaningful. These are two very well-known methods for root finding and combining the two of them is pretty common.

jhkolb replied:
Cool, looks good then.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/LoadBalancer.scala:
```diff
@@ -0,0 +1,323 @@
+  }
+
+  /**
+    * Compares the relavite location of two input centroids
```
relavite => relative, I think?

shellshocked2003 replied:
What are you talgink about?

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -29,12 +29,15 @@ object RockSlicer {
-    val (seedJoints, remainingJoints) = generateSeedJoints(joints, arguments.numSeedJoints)
-    seedJoints foreach { joint =>
-      blocks = blocks.flatMap(_.cut(joint, arguments.minRadius, arguments.maxAspectRatio))
+    if (arguments.numSeedJoints > 0) {
```
I'm kind of thinking we can get rid of the seed joints idea and just use processor joints to give a region of the rock volume to each processor. Seems like this would simplify things, since it's one less parameter the user has to understand. What do you think?

shellshocked2003 replied:
Seed joints are now a thing of the past

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -47,14 +50,65 @@ object RockSlicer {
-    val centroidBlocks = nonRedundantBlocks.map { block =>
+    // Find all blocks that contain processor joints
+    val processorBlocks = nonRedundantBlocks.filter { block => 
+      block.faces.exists { face => face.processorJoint}
```
What about `block.faces.exists(_.processorJoint)`?

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -47,14 +50,65 @@ object RockSlicer {
+    // Find blocks that do not contain processor joints
+    val realBlocks = nonRedundantBlocks.filter { block =>
+      val faceFlags = block.faces map { face =>
+        face.processorJoint
```
You should use `block.faces.any(_.processorJoint)`.

shellshocked2003 replied:
I think `any` is Haskell, but I'll assume you meant `exists` :)

jhkolb replied:
Yep, and I think Python too.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -47,14 +50,65 @@ object RockSlicer {
+        face.processorJoint
+      }
+      !faceFlags.contains(true)
+    }
```
Actually, Scala should have a `partition` method for `Seq` that can split it into these two collections for you, and it will be clearer and more concise that way.

shellshocked2003 replied:
`Partition` doesn't work on RDD's, so it can't be used in this instance.

jhkolb replied:
Aha, I see. I was just reading that if we want to filter the same RDD twice, we may want to take advantage of some of Spark's caching features. Another thing for us to look into I guess.

shellshocked2003 replied:
Yeah, roger that. Looks like there's a lot to look into :)

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -47,14 +50,65 @@ object RockSlicer {
+      !faceFlags.contains(true)
+    }
+
+    // PROCESS PROCESSOR BLOCKS BEFORE CONTINUEING WITH REMAINING BLOCKS
```
What's up with the all caps? :)

shellshocked2003 replied:
I was typing angry

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -47,14 +50,65 @@ object RockSlicer {
+    val allProcessorBlocks = processorBlocks.collect()
+    // Search blocks for matching processor joints
+    if (allProcessorBlocks.length > 0) {
+      val reconstructedBlocks = mergeBlocks(allProcessorBlocks, Seq[Block](), globalOrigin)
```
How about `Seq.empty[Block]` here?

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -47,14 +50,65 @@ object RockSlicer {
+
+      // Convert reconstructed blocks to requested output
+      if (arguments.toInequalities) {
+        val jsonReconBlocks = squeakyCleanRecon.map(Json.blockToMinimalJson)
```
It seems weird to me that we're separating the reconstructed blocks into their own file. Won't this be inconvenient for the user?

I think we should keep everything in a single file and continue to use Spark's text file utilities.

shellshocked2003 replied:
The reason I separated them into their own file is because all the "processor" have been collected onto one node. Spark's text file utilities don't work in this instance. 
I was thinking about how we can have all the nodes participate in this step instead of having only one node do all this work. If we do that, we can use Spark's text file utilities. I think grouping the blocks based on them containing the same processor joint and then parallelizing the collection to merge them will be a better approach. Currently looking into this.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +130,129 @@ object RockSlicer {
+          // Check for any actual shared faces between blocks - if these exists blocks
+          // should not be merged since there is a real joint seperating the blocks
+          val nonSharedFaces =
+            allFaces.foldLeft(Seq[Face]()) { (unique, current) =>
```
Perhaps `Seq.empty[Face]`

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +130,129 @@ object RockSlicer {
+          }
+        }
+      }
+    }).collect { case blockType: Block => blockType }
```
This seems weird. Can you explain why it's necessary? I think I have an idea, but I want to see what you say.

If I'm right, then this `collect` is unnecessary. You could just replace the `map` on line 149 with a `collect` I think.

shellshocked2003 replied:
In some cases there are no matches between blocks, so it returns an empty list. This gets rid of those empty entries, so I think what you're thinking is probably right.

jhkolb replied:
Cool. So I think you might be able to use a `collect` in place of your `map` then.

shellshocked2003 replied:
Tried what you suggested, but I don't think it'll work. `collect` doesn't seem to eliminate the empty lists. I think because it's being called too early and everything at that point is of `Block` type, nothing gets filtered out. When `collect` it called after all the calcs have been done, the empty lists are removed.

jhkolb replied:
Okay, what about taking the obvious approach and just filtering for non-empty lists? I think that would be better because `collect`, if I'm understanding right, is really meant to be used with partial functions.

shellshocked2003 replied:
The issue with filtering for non-empty lists is that `nonEmpty` is not a member of `Any`. I tried that before, but the issue is that the return type is `Unit` when it doesn't find any blocks that match. That's how I ended up using `collect` with a partial function. I could add an `else` that returns an empty list if you think that will make more sense.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +130,129 @@ object RockSlicer {
     }
-  }
+    val joinedBlocks = 
+      nonRedundantBlocks.foldLeft(Seq[Block]()) { (unique, current) =>
```
`Seq[Block].empty`

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +130,129 @@ object RockSlicer {
-        }
-        median +: loadBalanceIndicesImpl(queue)
+    // Divide blocks into groups with and without processor joints remaining
+    val remainingBlocks = joinedBlocks filter { block =>
```
Like before, you can use `partition` here to make this easier.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +130,129 @@ object RockSlicer {
+    val faceMatches = 
+      processorFaces1 map { case face1 =>
+        processorFaces2 map { case face2 =>
+          if ((math.abs(face1.a + face2.a) < NumericUtils.EPSILON) &&
```
Is there a reason you don't just use `compareSharedFaces` here?

shellshocked2003 replied:
No - it was written later and forgot to incorporate it here.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +130,129 @@ object RockSlicer {
-    loadBalanceIndicesImpl(queue)
+  // Function that writes JSON string to file for single node - taken from 
+  // http://stackoverflow.com/questions/4604237/how-to-write-to-a-file-in-scala
+  private def printToFile(f: java.io.File)(op: java.io.PrintWriter => Unit) {
```
Hmm, this code is pretty weird to me. There are more straightforward ways to accomplish this, but I actually think we should just stick to Spark's utilities anyways.

shellshocked2003 replied:
See the earlier comment about Spark's text utilities.

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/EndToEndSpec.scala:
```diff
@@ -17,26 +17,57 @@ class EndToEndSpec extends FunSuite {
-    // Calculate the centroid of each block
-    val centroidBlocks = nonRedundantBlocks.map { block =>
+    // Find all blocks that contain processor joints
+    val processorBlocks = nonRedundantBlocks.filter { block => 
```
Use Scala's `partition` method here.

shellshocked2003 replied:
Would it be better to leave it as-is since this is how the code needs to be for Spark?

jhkolb replied:
Maybe. I think it would be okay to use `partition`, but if we stick with the two calls to `filter` let's add a comment explaining that we do this to match the Spark code.

shellshocked2003 replied:
Okay. I added a comment about this. If anything, it'll serve as a reminder to us about why there's two filter statements.

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/RockSlicerSpec.scala:
```diff
@@ -0,0 +1,361 @@
+  test("Two blocks should be merged into unit cube") {
+    val globalOrigin = (0.0, 0.0, 0.0)
+    val processorBlocks = Seq(leftHalfUnitCube, rightHalfUnitCube)
+    val mergedBlocks = RockSlicer.mergeBlocks(processorBlocks, Seq[Block](), globalOrigin)
```
How about `Seq.empty[Block]`, and similar for the other cases in this file?

jhkolb commented:
First round of comments are done. Take a look.

shellshocked2003 commented:
Addressed most of your comments. I'll change all the `seed joints` stuff into `processor joints` tomorrow!

shellshocked2003 commented:
Got rid of the whole `seed joint` concept - everything is now in terms of number of processors, so users don't have to worry about what a `seed joint` is.
I think your comments have all been addressed, except for questions on two or so of them. Take a look and let me know what you think.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/Block.scala:
```diff
@@ -335,8 +394,10 @@ case class Block(center: (Double,Double,Double), faces: Seq[Face]) {
-        // If z-component of normal is negative, order needs to be reversed to maintain clockwise orientation
-        if (face.c < -NumericUtils.EPSILON) {
+        // If normal is -z-axis, order needs to be reversed to maintain clockwise orientation since
+        // rotation matrix is identity matrix in this case - vectors are parallel. Rotation matrix
```
Didn't you say this was when the vectors are pointing in exactly opposite directions, or something like that?

shellshocked2003 replied:
The rotation matrix will be the identity when vectors are parallel, whether they point in opposite directions or the same direction. The only time we need to change the order is when the normal is the -z-axis since the ordering will end up being clockwise instead of counterclockwise. We're rotating all points so they lie in the z-plane - this allows us to do a 2-D triangulation instead of 3-D. The two special cases where the triangulation will use the points as they are without rotating them is when the normal of the plane is the +/- z-axis. For the +z-axis, everything is fine, but we need to reverse the order for the -z-axis since the points will be ordered as if the normal was the +z-axis. 
Does that make sense? I can add more comments about this, but not sure how much detail is necessary.

jhkolb replied:
Okay, yeah I think that makes sense.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +130,120 @@ object RockSlicer {
+        processorFaces2 map { case face2 =>
+          if (compareSharedFaces(face1, face2)) {
+            Seq[Face](face1, face2)
+          } else Seq[Face]()
```
This `else` needs braces, and how about `Seq.empty[Face]`?

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +130,120 @@ object RockSlicer {
-        median +: loadBalanceIndicesImpl(queue)
       }
-    }
+    faceMatches.filter{ case faces => faces.nonEmpty }.flatten.flatten.distinct
```
You can just do `faceMatches.filter(_.nonEmpty)` here, right? You don't have to use the whole `case` syntax in some simple situations, and in fact you don't really need `case` unless you want to destructure each element, so `faceMatches.filter { faces => faces.nonEmpty }` should work here.

shellshocked2003 replied:
Yeah, I'm just starting to get that now. This branch is littered with unnecessary `case`s; hopefully they're gone now...

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/EndToEndSpec.scala:
```diff
@@ -17,26 +17,61 @@ class EndToEndSpec extends FunSuite {
+
+    // Find blocks that do not contain processor joints
+    val realBlocks = nonRedundantBlocks.filter { block =>
+      val faceTests = block.faces map { face =>
```
How about changing this to use `exists` like in your other code?

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/RockSlicerSpec.scala:
```diff
@@ -0,0 +1,361 @@
+    }
+
+    assert(mergedBlocks.length == 3)
+    assert(!blockCheck.contains(false))
```
Probably just want to use a `forAll` here.

jhkolb commented on src/test/scala/edu/berkeley/ce/rockslicing/RockSlicerSpec.scala:
```diff
@@ -0,0 +1,361 @@
+    }
+
+    assert(mergedBlocks.length == 3)
+    assert(!blockCheck.contains(false))
```
`forall`

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +130,120 @@ object RockSlicer {
+            allFaces.foldLeft(Seq.empty[Face]) { (unique, current) =>
+              if (!unique.exists(compareSharedFaces(current, _)))
+                current +: unique
+              else unique
```
Need braces here.

shellshocked2003 replied:
I thought I was being cute not using braces here, but I guess I've violated one of the main principles of defensive coding...

jhkolb commented:
Alright, next round of feedback is done. Take a look.

shellshocked2003 commented:
I think I've addressed all of your comments. Take a look and let me know what you think.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +128,126 @@ object RockSlicer {
+      // All blocks are free of processor joints - check for duplicates then return
+      val mergedBlocksDuplicates = completedBlocks ++ mergedBlocks
+      mergedBlocksDuplicates.foldLeft(Seq[Block]()) { (unique, current) =>
+        if (!unique.exists(current.approximateEquals(_)))
```
Here's another spot where we need braces.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +128,126 @@ object RockSlicer {
-        if (right.nonEmpty) {
-          queue.enqueue(right)
+      }
+    }).collect { case blockType: Block => blockType }
```
Okay, I think I see what you're saying. Let's talk about this when we meet tomorrow, I think I will be able to make better suggestions once you can walk me through this in person.

This seems like it's an important aspect of the implementation, so I think it's better we clean it up now to save ourselves trouble in the future.

jhkolb commented:
Okay, I think this branch is very close to ready. If you don't mind, let's just run through this one last time at the beginning of our meeting tomorrow before we merge, but hopefully that won't take too long.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -47,14 +50,63 @@ object RockSlicer {
+      }
+    }
+
+    // PROCESS REMAINING BLOCKS THAT DO NOT CONTAIN PROCESSOR BLOCKS
```
No caps here.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +128,127 @@ object RockSlicer {
+  @tailrec
+  def mergeBlocks(processorBlocks: Seq[Block], mergedBlocks: Seq[Block],
+                  origin: (Double, Double, Double)): Seq[Block] = {
+    val reconstructedBlocks = (processorBlocks map { block =>
```
You can make this `processorBlocks.tail map ...` to avoid duplicate comparisons.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +128,127 @@ object RockSlicer {
+          // should not be merged since there is a real joint seperating the blocks
+          val nonSharedFaces =
+            allFaces.foldLeft(Seq.empty[Face]) { (unique, current) =>
+              if (!unique.exists(compareSharedFaces(current, _))) {
```
How about moving the `comparedSharedFaces` function inside the `Face` class and calling it something like `isSharedWith` ? Then we can have code that looks like `face1.isSharedWith(face2)`.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +128,127 @@ object RockSlicer {
+                unique
+              }
+            }
+          if (allFaces.diff(nonSharedFaces).isEmpty) {
```
How about returning `Some(Block ...)` in the `if` and `None` in the `else`? Then just flatten at the end instead of collecting.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +128,127 @@ object RockSlicer {
+    val nonRedundantBlocks = reconstructedBlocks map { case block @ Block(center, _) =>
+      Block(center, block.nonRedundantFaces)
+    }
+    val joinedBlocks = 
```
We seem to think this isn't necessary anymore.

shellshocked2003 replied:
Yep, this isn't necessary.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +128,127 @@ object RockSlicer {
+    } else if (processorBlocks.tail.isEmpty) {
+      // All blocks are free of processor joints - check for duplicates then return
+      val mergedBlocksDuplicates = completedBlocks ++ mergedBlocks
+      mergedBlocksDuplicates.foldLeft(Seq[Block]()) { (unique, current) =>
```
You can remove this duplicate checking.

shellshocked2003 replied:
This duplicate checking needs to stay -  when `remainingBlocks` gets prepended to `processorBlocks.tail` there is the possibility that the same blocks can be generated through different merging orders.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +128,127 @@ object RockSlicer {
+  def compareProcessorBlocks(block1: Block, block2: Block): Seq[Face] = {
+    val processorFaces1 = block1.faces.filter { case face => face.processorJoint }
+    val processorFaces2 = block2.faces.filter { case face => face.processorJoint }
+    val faceMatches = 
```
You can probably just sort these as you did before. And you can drop the `case`s.

Also, instead of doing a bunch of separate sort operations, let's write a comparison function that does what we want with just one sort operation. Also should do this for the earlier face comparisons.

shellshocked2003 replied:
So, I tried implementing a comparison function, but it doesn't seem to be working with `sortWith`. Do you have any idea why? Seems like it should be working... Here's what it looks like:

``` scala
def faceSort(inputFace: Face): Boolean = {
    (d < inputFace.d) || (a < inputFace.a) || (b < inputFace.b) || (c <= inputFace.c)
}
```

Then I just do `faces.sortWith(_.faceSort(_))`, which doesn't give the correct ordering.

shellshocked2003 replied:
Also at Line 220, I think we'll need to keep it as-is since the faces aren't equal. We're looking for normal vectors and distances that are opposite, so sorting won't give the desired effect.
The nested `map` isn't done over all faces, only those containing a processor joint so it isn't quite as bad as we were thinking... Each block should contain only 2 to 3 processor joints at max.

shellshocked2003 commented:
Made some updates based on our discussion, but check out my comments about the sorting. Also, fixed a bug in `RockSlicer.mergeBlocks`. Some blocks share a processor joint, but actually are not anywhere close to each other. When calling `redundantFaces` on these merged blocks, they'd end up with an empty list for the faces. Changed it so these blocks get filtered out.

jhkolb commented:
I found some good info on sorting: http://stackoverflow.com/questions/19345030/easy-idiomatic-way-to-define-ordering-for-a-simple-case-class.

Let's go ahead and use implicits. So now instead of all those `sortWiths` you do, you can try something like this:

``` scala
faces.sortBy(face => (face.d, face.a, face.c, face.d))
```

jhkolb commented:
Okay, the changes generally look good. Let's fix up the sorting issue, I'll do one last look, and we'll merge.

shellshocked2003 commented:
The implicit ordering on `Tuple`s is pretty cool! Implemented the sort you suggested based on that. You're okay with keeping the nested `map` on Line 220 of `RockSlicer.scala`? Other than that, hopefully it's good to go.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +127,107 @@ object RockSlicer {
+    *         empty if they share no faces
+    */
+  def compareProcessorBlocks(block1: Block, block2: Block): Seq[Face] = {
+    val processorFaces1 = block1.faces.filter { face => face.processorJoint }
```
Can you just do `_.processorJoint` here?

shellshocked2003 replied:
Yeah, sorry I missed this.

jhkolb commented on src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala:
```diff
@@ -75,49 +127,107 @@ object RockSlicer {
+    val processorFaces2 = block2.faces.filter { face => face.processorJoint }
+    val faceMatches = 
+      processorFaces1 map { case face1 =>
+        processorFaces2 map { case face2 =>
```
Let's think carefully about if these nested `map`s are really necessary. If we're looking for faces with equal and opposite normal vectors, what about using a sort function that involves absolute value?

shellshocked2003 replied:
I think the primary reason this won't work is because the number of processor joints per block aren't guaranteed to be the same across all blocks, so even when you do a sort how do you do a fair comparison between the faces without some sort of map? I tried implementing a sorting-type solution, but it only works when the number of processor joints per block are the same between the two blocks being compared. 
For example, if you have a single processor joint in one block and three in another, sorting is somewhat pointless since you'll need to check the single joint against each of the other three for a match anyhow.
Do you agree with that?

jhkolb commented:
Couple more comments.

shellshocked2003 commented:
Couple more responses.

shellshocked2003 commented:
Made a minor addition to `Block.approximateEquals` - if length of `faces` is different, blocks are not equal.

jhkolb commented:
Merging. Yay!
